routers/repo/branch.go

```diff
@@ -195,100 +216,129 @@ func loadBranches(ctx *context.Context) []*Branch {
 	repoIDToGitRepo := map[int64]*git.Repository{}
 	repoIDToGitRepo[ctx.Repo.Repository.ID] = ctx.Repo.GitRepo
 
-	branches := make([]*Branch, len(rawBranches))
-	for i := range rawBranches {
-		commit, err := rawBranches[i].GetCommit()
+	var totalNumOfBranches = len(rawBranches)
+	var startIndex = (page - 1) * pageSize
+	if startIndex > totalNumOfBranches {
+		startIndex = totalNumOfBranches - 1
+	}
+	var endIndex = startIndex + pageSize
+	if endIndex > totalNumOfBranches {
+		endIndex = totalNumOfBranches - 1
+	}
+
+	var branches []*Branch
+	for i := startIndex; i < endIndex; i++ {
+		var branch = loadOneBranch(ctx, rawBranches[i], protectedBranches, repoIDToRepo, repoIDToGitRepo)
+		if branch == nil {
+			return nil, 0
+		}
+
+		if branch.Name == ctx.Repo.Repository.DefaultBranch {
```

Comment:
then 
```
if endIndex < totalNumOfBranches {
  endIndex++
}
```
Otherwise, we will get 19 branches this page.

Reply:
That’s actually my original logic. But I found that if the current page is the last page, and ‘master’ is also in this page, adding `endIndex++` will make the index out of bound. 

In fact, the entire pager mechanism is still based on the assumption that ‘master’ in the list. If we need to remove master from the list, the startIndex calculation also needs to be modified. I did try to change this, but it turns out to be too much work. I decided to left that for future work.

Reply:
That's why I added `if endIndex < totalNumOfBranches {`.

The default branch will always be kept on the top of the branches list table. That make the user feel, the paging is only for the list table but not all the branches on the page. So that, I think keep 20 branches on the table but not 19 is less confusing.

Reply:
Let me add some tests to clarify this. Could you please point me where i should be adding test? Does this affect the public documentation? I think there is still a lot work that needs to be done.

Reply:
You can add some tests on `integrations/branches_test.go`

Reply:
it might be easier to just always have the default branch on the page in a separate block.

that way you could simplify the logic.

Reply:
I tried to add some mock repos. It makes the PR really big, though. And I still get 404 instead of 200. Anything I'm missing? @lunny  @zeripath .

Reply:
OK I reverted the "Mock data" PR. The problem with `endIndex++` is that I can not just skip 'master' without breaking the content on the next page. Say we have 60 branches (including 'master'). If we do not need 'master' along, we have this:

```
Page0:  Branch0  ...  Branch19
Page1:  Branch20 ... master ... Branch39
Page2: Branch40 ...  Branch59
```

We need 'master' along and not included in the branches below, then we have this:

```
Page0:  Branch0  ...  Branch19
Page1:  Branch20 ... Branch39 (There is only 19 branches)
Page2: Branch40 ...  Branch59
```

This is the current solution. @lunny I'm using `for i := startIndex; i < endIndex; i++` so there is indeed 20 branches instead of 19 branches. Ideally we should be doing this:

```
Page0:  Branch0  ...  Branch19
Page1:  Branch20 ... Branch40
Page2: Branch41 ...  Branch59 (How do I know this should start at 41?)
```

But I cannot know that the third page should start at Branch41 unless I loop through all branches and find out where 'master' is, which seems inefficient.

Reply:
Or we can still display default branch in the list but give it a mark that means it's default.

Reply:
Could we don't ignore the default branch but mark it?